### ISSUE TITLE
Reference leak affecting all fusefs dentries

### ISSUE THREAD

**Issue by dean-deng** [google/gvisor#4074]
Fusefs is built on top of kernfs, which currently does not release references on dentries obtained during Lookup() until they are invalidated (see kernfs.Filesystem.revalidateChildLocked). Fusefs dentries are never invalidated.

**Comment by ayushr2** [comment]
Yeah I have been looking at this. I will be addressing this in one of my upcoming changes. The issue is that currently all Inodes that inherit `InodeNotDirectory` are always valid by default: https://github.com/google/gvisor/blob/0a7075f38a4870ded687e117a299ac4996c0673e/pkg/sentry/fsimpl/kernfs/inode_impl_util.go#L149-L151

Even `/proc/[pid]/fd{info}/[fd]` inodes are always valid when they should be able to be invalidated.

**Comment by ayushr2** [comment]
As discussed with @mrahatm offline, the fuse bug is a different one. Unassigning myself

**Comment by mrahatm** [comment]
This one is slightly different from the procfs case. Here, the real bug is fusefs shouldn't implement dynamic dentries (i.e. Lookup() should always return nil).

**Comment by ayushr2** [comment]
@manninglucas do you think this is still the case?

**Comment by manninglucas** [comment]
I think it's still the case (although I don't have context on the offline discussion). The kernfs-based dentries always return that they are valid and `Keep()` always returns true, so they're never revalidated/released.
